lib/maths/analytics/COutliers.cc

```diff
@@ -600,79 +601,68 @@ void CEnsemble<POINT>::CScorer::add(const TMeanVarAccumulator2Vec& logScoreMomen
         return 0.5;
     };
 
-    auto pOutlierGiven = [](double cdfComplement) {
-
+    auto pOutlierGiven = [] {
         static const TDoubleVec LOG_KNOTS{
-            common::CTools::fastLog(1e-5), common::CTools::fastLog(1e-3),
-            common::CTools::fastLog(0.01), common::CTools::fastLog(0.1),
-            common::CTools::fastLog(0.2),  common::CTools::fastLog(0.4),
-            common::CTools::fastLog(0.5),  common::CTools::fastLog(0.6),
-            common::CTools::fastLog(0.8),  common::CTools::fastLog(1.0)};
-        static const TDoubleVec KNOTS_P_OUTLIER{0.98, 0.87, 0.76, 0.65, 0.6,
-                                                0.5,  0.5,  0.5,  0.3,  0.3};
-
-        double logCdfComplement{common::CTools::fastLog(std::max(cdfComplement, 1e-5))};
-        auto k = std::upper_bound(LOG_KNOTS.begin(), LOG_KNOTS.end(), logCdfComplement);
-
-        return common::CTools::linearlyInterpolate(
-            *(k - 1), *k, KNOTS_P_OUTLIER[k - LOG_KNOTS.begin() - 1],
-            KNOTS_P_OUTLIER[k - LOG_KNOTS.begin()], logCdfComplement);
-    };
+            common::CTools::fastLog(std::numeric_limits<double>::min()),
+            common::CTools::fastLog(1e-5),
+            common::CTools::fastLog(1e-3),
+            common::CTools::fastLog(0.01),
+            common::CTools::fastLog(0.1),
+            common::CTools::fastLog(0.2),
+            common::CTools::fastLog(0.4),
+            common::CTools::fastLog(0.5),
+            common::CTools::fastLog(0.6),
+            common::CTools::fastLog(0.8),
+            common::CTools::fastLog(1.0)};
+        static const TDoubleVec KNOTS_P_OUTLIER{
+            0.9999, 0.99, 0.96, 0.82, 0.67, 0.62, 0.51, 0.5, 0.5, 0.35, 0.33};
+        static common::CSpline<> P_OUTLIER{[&] {
+            common::CSpline<> result{common::CSplineTypes::E_Linear};
+            result.interpolate(LOG_KNOTS, KNOTS_P_OUTLIER,
+                               common::CSplineTypes::E_ParabolicRunout);
+            return result;
+        }()};
+
+        return [&](double cdfComplement) {
+            double logCdfComplement{common::CTools::fastLog(
+                std::max(cdfComplement, std::numeric_limits<double>::min()))};
+            return P_OUTLIER.value(logCdfComplement);
+        };
```

Comment:
Note that this represents a small change to how we compute outlier scores for extreme outliers. Before things would saturate which means we couldn't reliably compute the derivative of the overall score w.r.t. changes in feature values. This extends the range to `std::numeric_limits<double>::min()`. It also avoids making the function as flat around P = 0.5 for the same reason. The resulting scores are still broadly similar.